text/0000-system-receptionist.md

```diff
@@ -0,0 +1,255 @@
+
+The `system` value is made available in the 3rd argument to action/guard/invoke creators, which contains the `system.receptionist` that extends the `ActorRef` interface.
+
+Listing all `ActorRef` objects registered under a key is done via `system.receptionist.list(key)`, which synchronously returns an array of `ActorRef` objects.
+
+```ts
+const todosMachine = createMachine({
+  // ...
+  on: {
+    'todos.logAll': {
+      actions: (ctx, e, { system }) => {
+        // Lists all todo ActorRefs
+        console.log(system.receptionist.list('todo'));
+      }
+    },
+    'todos.completeAll': {
+      actions: pure((ctx, e, { system }) => {
+        return system.receptionist.list('todo').map(todoRef => {
+          return sendTo(todoRef, { type: 'complete' });
+        });
+      });
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.list('todo');
+system.receptionist.getSnapshot().registry.get('todo') ?? []; // => ActorRef[]
+```
+
+**Single listing**
+
+A common use-case of the receptionist will be to register and find a single `ActorRef` registered to a given `key`. This can be done via `system.registry.find(key)`, which will return `ActorRef | null`.
+
+```ts
+const todoMachine = createMachine({
+  // ...
+  on: {
+    complete: {
+      // `system.registry.find(key)` returns 0 or 1 ActorRef objects
+      // TODO: ensure `sendTo(...)` handles null targets (dead-letter queue?)
+      actions: sendTo((ctx, e, { system }) => system.receptionist.find('notifier'), {
+        type: 'notify',
+        message: 'Todo completed'
+      })
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.get('notifier');
+system.receptionist.getSnapshot().registry.get('notifier')?.[0] ?? undefined; // => ActorRef[]
+```
+
+**Subscribing to listings**
+
+The registry may change over time, as actors get registered and unregistered from certain keys. The `system.receptionist.listener(key)` returns an `ActorRef` (subscribable by nature) which can be subscribed to for the specific purpose of listening to changes in `ActorRef` objects registered under a key. The emitted value is the complete `ActorRef[]` array.
+
+```ts
+const todosMachine = createMachine({
+  invoke: {
+    src: (ctx, e, { system }) => fromObservable(() => system.receptionist.listener('todo')),
+    onEmit: {
+      actions: (_, event) => {
+        // ActorRef[] of 'todo' actor refs
+        console.log(event.data); 
+      }
+    }
+  }
+});
+```
+
+This should be roughly equivalent to:
+```ts
+// PSEUDOCODE
+// system.receptionist.listener('todo')
+const receptionist = {
+  onRegister(key, actorRef) {
+    // Register actorRef
+    // ...
+
+    observers[key].forEach(observer => {
+      observer.next(registry.get(key))
+    });
+  },
+
+  listener(key) {
+    // Return ActorRef that listens for registry updates on that key
+    // ...
+  }
+}
+```
+
+
+**Unregistering**
```

Comment:
We should also automatically clean up actors in the receptionist when an actors stops